File: prelude/1.0.0/js/language_runtime.js
class Atom {
  constructor(value) {
    this.value = value;
  }
}

export function atom(value) {
  if (arguments.length !== 1) throw new Error("atom expects one value");
  return new Atom(value);
}

export function deref(reference) {
  if (arguments.length !== 1 || !(reference instanceof Atom))
    throw new Error("deref expects one atom");
  return reference.value;
}

export function reset_BANG_(reference, value) {
  if (arguments.length !== 2 || !(reference instanceof Atom))
    throw new Error("reset! expects an atom and a value");
  reference.value = value;
  return value;
}

export function swap_BANG_(reference, fn) {
  if (arguments.length !== 2 || !(reference instanceof Atom) || typeof fn !== "function")
    throw new Error("swap! expects an atom and a function");
  // ponytail: sequential updates; synchronization needs a separate concurrency contract.
  const value = fn(reference.value);
  reference.value = value;
  return value;
}

function valueText(value) {
  if (value === null || value === undefined) return "nil";
  if (value instanceof Atom) return "#<atom>";
  if (Array.isArray(value)) return `(${value.map(valueText).join(" ")})`;
  if (typeof value === "object") {
    return `{${Object.entries(value)
      .map(([key, value]) => `${key} ${valueText(value)}`)
      .join(" ")}}`;
  }
  if (typeof value === "function") return "#<function>";
  return String(value);
}

function strText(value) {
  return valueText(value);
}

export function list(...items) {
  return items;
}

export function vector_QMARK_(value) {
  return Array.isArray(value);
}

export function _EQ_(left, right) {
  return Object.is(left, right);
}

export function not_EQ_(left, right) {
  return !_EQ_(left, right);
}

export function _GT_(left, right) {
  return left > right;
}

export function _LT_(left, right) {
  return left < right;
}

export function _GT__EQ_(left, right) {
  return left >= right;
}

export function _LT__EQ_(left, right) {
  return left <= right;
}

export function concat(...collections) {
  const result = [];
  for (const collection of collections) {
    if (!Array.isArray(collection)) throw new Error("concat expects lists");
    result.push(...collection);
  }
  return result;
}

export function hash_map(...items) {
  if (items.length % 2 !== 0) {
    throw new Error("hash-map arguments must be key/value pairs");
  }

  const result = Object.create(null);
  for (let i = 0; i < items.length; i += 2) {
    result[valueText(items[i])] = items[i + 1];
  }
  return result;
}

export function get(collection, key) {
  if (collection === null) return null;
  if (Array.isArray(collection)) return Number.isInteger(key) && key >= 0 ? collection[key] ?? null : null;
  if (collection !== null && typeof collection === "object") return collection[valueText(key)] ?? null;
  throw new Error("get expects a hash-map/list and a key/index");
}

export function get_in(collection, keys) {
  if (!Array.isArray(keys)) throw new Error("get-in expects a collection and a vector path");
  return keys.reduce((value, key) => get(value, key), collection);
}

export function truthy(value) {
  return (
    value !== false &&
    value !== null &&
    value !== undefined &&
    value !== "false" &&
    value !== "nil"
  );
}

export function not(value) {
  return !truthy(value);
}

export function print_result(value) {
  console.log(valueText(value));
}

export function println(...items) {
  console.log(items.map(strText).join(" "));
}

export function eprintln(...items) {
  console.error(items.map(strText).join(" "));
}

export function str(...items) {
  return items.map(strText).join("");
}

export function _PLUS_(...items) {
  const result = items.reduce((sum, item) => sum + item, 0);
  return result === 0 ? 0 : result;
}

export function _MINUS_(...items) {
  if (items.length === 0) throw new Error("- expects at least one number");
  const result = items.slice(1).reduce((result, item) => result - item, items[0]);
  return result === 0 ? 0 : result;
}

export function _STAR_(...items) {
  const result = items.reduce((result, item) => result * item, 1);
  return result === 0 ? 0 : result;
}

export function _SLASH_(...items) {
  if (items.length === 0) throw new Error("/ expects at least one number");
  return items.slice(1).reduce((result, item) => Math.trunc(result / item), items[0]);
}

export function count(collection) {
  if (Array.isArray(collection)) return collection.length;
  if (collection !== null && typeof collection === "object") {
    return Object.keys(collection).length;
  }
  throw new Error("count expects one collection");
}

export function map(fn, list) {
  if (!Array.isArray(list)) throw new Error("map expects a function and a list");
  return list.map((item) => fn(item));
}

export function drop(count, list) {
  return list.slice(Math.max(0, count));
}

export function reduce(fn, init, list) {
  const hasInit = arguments.length === 3;
  if (!hasInit) list = init;
  if (!Array.isArray(list)) {
    if (list !== null && typeof list === "object") list = Object.entries(list);
    else throw new Error("reduce expects a list or hash-map");
  }
  if (!hasInit && list.length === 0) throw new Error("reduce expects a non-empty list");
  return hasInit
    ? list.reduce((acc, item) => fn(acc, item), init)
    : list.reduce((acc, item) => fn(acc, item));
}

File: prelude/1.0.0/java/language_runtime.java
package y2k.language;

public final class language_runtime {

  @FunctionalInterface
  public interface Fn0 {
    Object call() throws Exception;
  }

  @FunctionalInterface
  public interface Fn1 {
    Object call(Object a) throws Exception;
  }

  @FunctionalInterface
  public interface Fn2 {
    Object call(Object a, Object b) throws Exception;
  }

  @FunctionalInterface
  public interface Fn3 {
    Object call(Object a, Object b, Object c) throws Exception;
  }

  @FunctionalInterface
  public interface Fn4 {
    Object call(Object a, Object b, Object c, Object d) throws Exception;
  }

  public static RuntimeException sneaky_throw(Throwable throwable) {
    language_runtime.<RuntimeException>sneaky_throw_unchecked(throwable);
    return null;
  }

  @SuppressWarnings("unchecked")
  private static <T extends Throwable> void sneaky_throw_unchecked(Throwable throwable) throws T {
    throw (T) throwable;
  }

  public static java.util.List<Object> list(Object... items) {
    return java.util.Arrays.asList(items);
  }

  public static Boolean vector_QMARK_(Object value) {
    return value instanceof java.util.List<?>;
  }

  public static java.util.List<Object> concat(Object... collections) {
    var result = new java.util.ArrayList<Object>();
    for (Object collection : collections) {
      if (!(collection instanceof java.util.List<?> items))
        throw new RuntimeException("concat expects lists");
      result.addAll(items);
    }
    return result;
  }

  public static java.util.LinkedHashMap<String, Object> hash_map(Object... items) {
    if (items.length % 2 != 0) {
      throw new RuntimeException("hash-map arguments must be key/value pairs");
    }
    var result = new java.util.LinkedHashMap<String, Object>();
    for (int i = 0; i < items.length; i += 2) {
      result.put(value_text(items[i]), items[i + 1]);
    }
    return result;
  }

  public static Object get(Object collection, Object key) {
    if (collection instanceof java.util.List<?> items && key instanceof Number index) {
      int i = index.intValue();
      return i >= 0 && i < items.size() ? items.get(i) : null;
    }
    if (collection instanceof java.util.Map<?, ?> items)
      return items.get(value_text(key));
    throw new RuntimeException("get expects a hash-map/list and a key/index");
  }

  public static boolean truthy(Object value) {
    return !(value == null || Boolean.FALSE.equals(value) || "false".equals(value) || "nil".equals(value));
  }

  public static Object print_result(Object value) {
    System.out.println(value_text(value));
    return null;
  }

  public static Object println(Object... items) {
    System.out.println(join_str(items, " "));
    return null;
  }

  public static Object eprintln(Object... items) {
    System.err.println(join_str(items, " "));
    return null;
  }

  public static Object missing(Object... items) {
    throw new RuntimeException("missing");
  }

  public static String str(Object... items) {
    return join_str(items, "");
  }

  public static Integer _PLUS_(Object... items) {
    int result = 0;
    for (Object item : items)
      result += ((Number) item).intValue();
    return result;
  }

  public static Integer _MINUS_(Object... items) {
    if (items.length == 0)
      throw new RuntimeException("- expects at least one number");
    int result = ((Number) items[0]).intValue();
    for (int i = 1; i < items.length; i++)
      result -= ((Number) items[i]).intValue();
    return result;
  }

  public static Integer _STAR_(Object... items) {
    int result = 1;
    for (Object item : items)
      result *= ((Number) item).intValue();
    return result;
  }

  public static Integer _SLASH_(Object... items) {
    if (items.length == 0)
      throw new RuntimeException("/ expects at least one number");
    int result = ((Number) items[0]).intValue();
    for (int i = 1; i < items.length; i++)
      result /= ((Number) items[i]).intValue();
    return result;
  }

  public static Integer count(Object collection) {
    if (collection instanceof java.util.List<?> list)
      return list.size();
    if (collection instanceof java.util.Map<?, ?> map)
      return map.size();
    throw new RuntimeException("count expects one collection");
  }

  public static java.util.List<Object> map(Object fn, Object collection) throws Exception {
    if (!(collection instanceof java.util.List<?> items)) {
      throw new RuntimeException("map expects a function and a list");
    }
    var result = new java.util.ArrayList<Object>();
    for (Object item : items)
      result.add(call_fn(fn, item));
    return result;
  }

  public static java.util.List<Object> drop(Object count, Object collection) {
    var items = (java.util.List<?>) collection;
    int start = Math.min(Math.max(((Number) count).intValue(), 0), items.size());
    return new java.util.ArrayList<Object>(items.subList(start, items.size()));
  }

  public static Object reduce(Object fn, Object collection) throws Exception {
    if (!(collection instanceof java.util.List<?> items)) {
      throw new RuntimeException("reduce expects a function and a list");
    }
    if (items.isEmpty()) {
      throw new RuntimeException("reduce expects a non-empty list");
    }
    Object acc = items.get(0);
    for (int i = 1; i < items.size(); i++)
      acc = call_fn(fn, acc, items.get(i));
    return acc;
  }

  public static Object reduce(Object fn, Object init, Object collection) throws Exception {
    if (!(collection instanceof java.util.List<?> items)) {
      throw new RuntimeException("reduce expects a function and a list");
    }
    Object acc = init;
    for (Object item : items)
      acc = call_fn(fn, acc, item);
    return acc;
  }

  static Object call_fn(Object fn, Object... args) throws Exception {
    if (fn instanceof Fn0 callable) {
      expect_args("function", args, 0);
      return callable.call();
    }
    if (fn instanceof Fn1 callable) {
      expect_args("function", args, 1);
      return callable.call(args[0]);
    }
    if (fn instanceof Fn2 callable) {
      expect_args("function", args, 2);
      return callable.call(args[0], args[1]);
    }
    if (fn instanceof Fn3 callable) {
      expect_args("function", args, 3);
      return callable.call(args[0], args[1], args[2]);
    }
    if (fn instanceof Fn4 callable) {
      expect_args("function", args, 4);
      return callable.call(args[0], args[1], args[2], args[3]);
    }
    throw new RuntimeException("value is not a function");
  }

  static void expect_args(String name, Object[] args, int count) {
    if (args.length != count) {
      throw new RuntimeException(name + " expects " + count + " arguments");
    }
  }

  static String value_text(Object value) {
    if (value == null)
      return "nil";
    if (value instanceof java.util.List<?> list) {
      var items = new java.util.ArrayList<String>();
      for (Object item : list)
        items.add(value_text(item));
      return "(" + String.join(" ", items) + ")";
    }
    if (value instanceof java.util.Map<?, ?> map) {
      var items = new java.util.ArrayList<String>();
      for (var entry : map.entrySet()) {
        items.add(entry.getKey() + " " + value_text(entry.getValue()));
      }
      return "{" + String.join(" ", items) + "}";
    }
    if (value instanceof Fn0 || value instanceof Fn1 || value instanceof Fn2 || value instanceof Fn3
        || value instanceof Fn4)
      return "#<function>";
    return String.valueOf(value);
  }

  static String str_text(Object value) {
    return value_text(value);
  }

  static String join_str(Object[] items, String separator) {
    var parts = new java.util.ArrayList<String>();
    for (Object item : items)
      parts.add(str_text(item));
    return String.join(separator, parts);
  }
}
